<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <title>D:\a\hazelcast-csharp-client\hazelcast-csharp-client\src\Hazelcast.Net\Clustering\RemoteExceptions.cs</title>
    <script type="text/javascript" src="../js/dotcover.sourceview.js"></script>
    <link rel="stylesheet" type="text/css" href="../css/dotcover.report.css" />
  </head>
  <body>
    <pre id="content" class="source-code">
// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections.Generic;
using System.Text;
using Hazelcast.Protocol;
using Hazelcast.Protocol.Models;

namespace Hazelcast.Clustering
{
    /// &lt;summary&gt;
    /// Maps remote exceptions to C# exceptions.
    /// &lt;/summary&gt;
    internal static class RemoteExceptions
    {
        private static readonly HashSet&lt;RemoteError&gt; RetryableExceptions = new HashSet&lt;RemoteError&gt;
        {
            RemoteError.CallerNotMember,
            RemoteError.MemberLeft,
            RemoteError.PartitionMigrating,
            RemoteError.RetryableHazelcast,
            RemoteError.RetryableIO,
            RemoteError.TargetNotMember,
            RemoteError.WrongTarget,
            RemoteError.TargetNotReplicaException,
            RemoteError.CannotReplicateException,
            RemoteError.HazelcastInstanceNotActive,

            // this one must *not* automatically retryable, because it requires more
            // checks on the client and message in order to decide whether to retry
            //RemoteError.TargetDisconnected,
        };

        /// &lt;summary&gt;
        /// Creates a C# exception that represents an exception that was thrown remotely on a server.
        /// &lt;/summary&gt;
        /// &lt;param name=&quot;memberId&quot;&gt;The unique identifier of the member that threw the exception.&lt;/param&gt;
        /// &lt;param name=&quot;errorHolders&quot;&gt;Server errors.&lt;/param&gt;
        /// &lt;returns&gt;An exception representing the specified server errors.&lt;/returns&gt;
        internal static RemoteException CreateException(Guid memberId, IEnumerable&lt;ErrorHolder&gt; errorHolders)
        {
            if (errorHolders == null) throw new ArgumentNullException(nameof(errorHolders));

            return CreateException(memberId, errorHolders.GetEnumerator());
        }

        /// &lt;summary&gt;
        /// Create a C# exception that represents an exception that was thrown remotely on a server.
        /// &lt;/summary&gt;
        /// &lt;param name=&quot;memberId&quot;&gt;The unique identifier of the member that threw the exception.&lt;/param&gt;
        /// &lt;param name=&quot;errorHolders&quot;&gt;Server errors.&lt;/param&gt;
        /// &lt;returns&gt;An exception representing the specified server errors.&lt;/returns&gt;
        private static RemoteException CreateException(Guid memberId, IEnumerator&lt;ErrorHolder&gt; errorHolders)
        {
            if (errorHolders == null) throw new ArgumentNullException(nameof(errorHolders));

            if (!errorHolders.MoveNext())
                return null;

            var errorHolder = errorHolders.Current;
            if (errorHolder == null) return new RemoteException(memberId, RemoteError.Undefined);

            var innerException = CreateException(memberId, errorHolders);

            var error = RemoteError.Undefined;
            if (Enum.IsDefined(typeof(RemoteError), errorHolder.ErrorCode))
                error = (RemoteError) errorHolder.ErrorCode;
            var retryable = RetryableExceptions.Contains(error);

            var serverStackTrace = new StringBuilder();
            var first = true;
            foreach (var stackTraceElement in errorHolder.StackTraceElements)
            {
                if (first) first = false; else serverStackTrace.AppendLine();
                serverStackTrace.Append(&quot;   &quot;).Append(stackTraceElement);
            }

            return new RemoteException(memberId, error, errorHolder.Message, innerException, serverStackTrace.ToString(), retryable);
        }
    }
}

    </pre>
    <script type="text/javascript">
      highlightRanges([[28,9,44,11,1],[54,13,54,38,1],[54,39,54,93,0],[56,13,56,76,1],[67,13,67,38,1],[67,39,67,93,0],[69,13,69,42,1],[70,17,70,29,1],[72,13,72,52,1],[73,13,73,37,1],[73,38,73,98,0],[75,13,75,74,1],[77,13,77,47,1],[78,13,78,76,1],[79,17,79,61,1],[80,13,80,65,1],[82,13,82,56,1],[83,13,83,30,1],[84,22,84,43,1],[84,44,84,46,1],[84,47,84,77,1],[86,17,86,27,1],[86,28,86,42,1],[86,48,86,78,1],[87,17,87,74,1],[90,13,90,134,1]]);
    </script>
  </body>
</html>